multilep/src/GenTools.cc

```diff
@@ -213,12 +217,50 @@ unsigned GenTools::provenanceConversion(const reco::GenParticle* photon, const s
     return 0;
 }
 
+
+/*
+ * Identify the leptons in the HNL samples. Matching l2 and l3 (2 and 3) does not work for MN < MW as the off-shell W
+ * not stored in the decay chain. Is used as a tool to mainly tag l1 in isDiracType
+ */
+unsigned GenTools::provenanceHeavyNeutrino(const reco::GenParticle& genParticle, const std::vector<reco::GenParticle>& genParticles){
+    //0: nonprompt 
+    //1: no HNL in chain
+    //2: HNL is mother
+    //3: HNL in chain but not mother
+    if(!isPrompt(genParticle, genParticles)) return 0; // This was how it was also defined in the old GenMatching code
+//    if(!gen) return 4;
+
+    std::set<int> decayChain;
+    setDecayChain(genParticle, genParticles, decayChain);
+    if(!heavyNeutrinoInChain(decayChain)){
+        return 1;
+    } else {
+        if(getMotherPdgId(genParticle, genParticles) == 9900012){
+            return 2;
+        } else {
+            return 3;
+        }
+    }
+    return 4;
+}
+
 bool GenTools::isPrompt(const reco::GenParticle& gen, const std::vector<reco::GenParticle>& genParticles){
     const reco::GenParticle* mom = getMother(gen, genParticles);
     if(abs(mom->pdgId()) == 15 && mom->isPromptDecayed()) return true;
     return (gen.isPromptFinalState() || gen.isPromptDecayed());
 }
 
+/*
+ * Check if the HNL sample is a Dirac type event or a LNV event
+ */
+bool GenTools::isDiracType(const std::vector<reco::GenParticle>& genParticles){
+    const reco::GenParticle* l1 = returnL1MatchForHNL(genParticles);
+    const reco::GenParticle* nu = returnNeutrinoMatchForHNL(genParticles);
+    if(l1->pdgId() > 0 && nu->pdgId() > 0) return false;
+    if(l1->pdgId() < 0 && nu->pdgId() < 0) return false;
```

Comment:
Can be simplified:

`return l1->pdgId() * nu->pdgId() < 0;`

Reply:
Done